### PR TITLE
Expose span locations on stable

### PR DESCRIPTION
**Pull request by dtolnay**
In the way suggested by https://github.com/alexcrichton/proc-macro2/pull/165#issuecomment-457032402.

### CODE REVIEW

File: src/fallback.rs [hunk 1]
```diff
@@ -1,5 +1,3 @@
-#![cfg_attr(not(procmacro2_semver_exempt), allow(dead_code))]
-
 #[cfg(procmacro2_semver_exempt)]
```
Comment (mitsuhiko) [inline src/fallback.rs]:
```suggestion
#[cfg(any(procmacro2_semver_exempt, feature = "span-locations"))]
```

Comment (mitsuhiko) [reply src/fallback.rs]: Needs a few more similar code paths further down.

**Comment by mitsuhiko** [comment]
Currently fails like this with the feature enabled:

```
error[E0433]: failed to resolve: use of undeclared type or module `RefCell`
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:228:40
    |
228 |     static CODEMAP: RefCell<Codemap> = RefCell::new(Codemap {
    |                                        ^^^^^^^ use of undeclared type or module `RefCell`

error[E0412]: cannot find type `RefCell` in this scope
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:228:21
    |
228 |     static CODEMAP: RefCell<Codemap> = RefCell::new(Codemap {
    |                     ^^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
    |
5   | use std::cell::RefCell;
    |

error[E0412]: cannot find type `Codemap` in this scope
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:228:29
    |
228 |     static CODEMAP: RefCell<Codemap> = RefCell::new(Codemap {
    |                             ^^^^^^^ not found in this scope

error[E0422]: cannot find struct, variant or union type `Codemap` in this scope
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:228:53
    |
228 |     static CODEMAP: RefCell<Codemap> = RefCell::new(Codemap {
    |                                                     ^^^^^^^ not found in this scope

error[E0422]: cannot find struct, variant or union type `FileInfo` in this scope
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:231:21
    |
231 |         files: vec![FileInfo {
    |                     ^^^^^^^^ not found in this scope

error[E0609]: no field `lo` on type `&fallback::Span`
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:379:40
    |
379 |             fi.offset_line_column(self.lo as usize)
    |                                        ^^

error[E0609]: no field `hi` on type `&fallback::Span`
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:388:40
    |
388 |             fi.offset_line_column(self.hi as usize)
    |                                        ^^

error[E0560]: struct `fallback::Span` has no field named `lo`
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:233:26
    |
233 |             span: Span { lo: 0, hi: 0 },
    |                          ^^ `fallback::Span` does not have this field

error[E0560]: struct `fallback::Span` has no field named `hi`
   --> /Users/mitsuhiko/.cargo/git/checkouts/proc-macro2-bc9077ffc3c20345/f24566c/src/fallback.rs:233:33
    |
233 |             span: Span { lo: 0, hi: 0 },
    |                                 ^^ `fallback::Span` does not have this field

error: aborting due to 9 previous errors
```

File: src/fallback.rs (hunk 1 above)
Comment (dtolnay) [reply src/fallback.rs]: Fixed.

**Comment by mitsuhiko** [comment]
Can confirm this works!

**Comment by dtolnay** [comment]
Tagging <kbd>wip</kbd> because this needs tests. Does not work when I tried it:

```rust
use proc_macro2::TokenTree;

fn main() {
    let sp = syn::parse_str::<TokenTree>("testing").unwrap().span();
    println!("start line={} column={}", sp.start().line, sp.start().column);
    println!("end line={} column={}", sp.end().line, sp.end().column);
}
```

```console
start line=1 column=0
end line=1 column=0
```

I am not going to be able to get to this today. @mitsuhiko could you investigate?

**Comment by mitsuhiko** [comment]
Yes, will investigate. Information seems absent indeed.

**Comment by alexcrichton** [comment]
If we're going to offer this in a published stable version then I think we can probably go ahead and do this without a feature. The `proc-macro2` crate can't have many major version bumps due to the degree of reliance on it through `syn` in the proc-macro ecosystem, so being stable is a big commitment. If we're willing to commit to this as a stable feature then I think it's worth providing by default.

I think that may also simplify the cfg story here (which I am always confused by every time I look at this!). That way outside of a proc-macro context you always have accurate information, but in a proc-macro context you'll only have accurate information if a nightly feature toggle is enabled.

**Comment by dtolnay** [comment]
I kept a feature because last time I checked there was a big performance hit for Span going from 0 bytes to 8 bytes. There are a lot of spans in the syntax tree...

**Comment by alexcrichton** [comment]
Ah ok that makes sense to me, and in that case seems fine to document as such as to why it's a feature

**Comment by mystor** [comment]
Hmm, I wonder if a span interning system could compress span sizes down even more, and if that would be helpful here.

**Comment by dtolnay** [comment]
I added a note in build.rs about the performance hit.

**Comment by dtolnay** [comment]
I think this works now.

I filed #167 to follow up on mitigating the performance impact.

**Comment by mitsuhiko** [comment]
Is this blocked on #167 or is it mergeable as such?

**Comment by dtolnay** [comment]
Published in 0.4.27.

**Comment by chinedufn** [comment]
@dtolnay noob question here - is this performance hit increased memory, noticeably slower speed for working with a TokenStream.. or both.. or something else?

#167 suggests that it's increased memory but I'm not sure if that's also having a big impact on other things?

Trying to better understand whether or not to start taking advantage of this.

Thanks!!

**Comment by dtolnay** [comment]
@chinedufn it was noticeably longer compile time in debug mode for Serde's test suite the last time I tried it. You could enable the feature and check the compile time for some crate that uses your macro heavily.

**Comment by chinedufn** [comment]
Awesome advice, perfect, thanks!!

**Comment by chinedufn** [comment]
Hmm so I'm seeing some weird behavior for span locations

Take the following proc macro

```
#[test]
fn text_space_before_end_tag() {
    assert_eq!(
        &html! { <div>Before End Tag </div> }.to_string(),
        "<div>Before End Tag </div>"
    )
}
```

For the token `Tag`, the span is perfectly correct. Its span's end column is 36, which .. after counting characters it's definitely the 36th column.

However, the `<` in `</div` has a start column of `37`.

I would expect it to be 38 here since there is a space between the end of the word `Tag` and the `<`.

Does that sound familiar at all? Held off on pushing up something easily reproducible in case there is an obvious reason here that you might already know but if not I can totally do that if that would help!

**Comment by alexcrichton** [comment]
@chinedufn that's likely actually a bug in rustc itself, want to see if you can reduce it to a standalone proc-macro and report it upstream?

**Comment by chinedufn** [comment]
Will do, thanks!

**Comment by chinedufn** [comment]
For posterity! -> https://github.com/rust-lang/rust/issues/58958

**Pull request merged**